packages/components/copybutton/src/CopyButton.tsx

```diff
@@ -44,76 +75,106 @@ export interface CopyButtonProps extends CommonProps {
     size = 'medium',
     ...otherProps
-  } = props;
-  const styles = getStyles({ size });
-
+  }: ExpandProps<CopyButtonProps>,
+  ref: React.Ref<HTMLDivElement>,
+) {
+  const styles = getCopyButtonStyles({ size });
+  const [isLoading, setIsLoading] = useState(false);
   const [copied, setCopied] = useState(false);
   const button = useRef<HTMLButtonElement | null>(null);
-  const handleOnCopy = useCallback(
-    (e: string) => {
-      if (onCopy) {
-        onCopy(e);
+  const resolvedValue = useRef<string | Promise<string>>(
+    typeof value === 'function' ? value() : value,
+  );
+  const timer = useRef<number | undefined>();
+
+  useEffect(() => {
+    return () => {
+      if (timer.current) {
+        clearTimeout(timer.current);
       }
+    };
+  }, []);
 
-      setCopied(true);
+  useEffect(() => {
+    const load = async () => {
+      if (preload && isPromiseLike(resolvedValue.current)) {
+        resolvedValue.current = await resolvedValue.current;
+      }
+    };
 
-      setTimeout(() => {
-        setCopied(false);
-        if (button.current) {
-          button.current.blur();
-        }
-      }, 1000);
-    },
-    [onCopy],
-  );
+    load();
+  }, [preload, value]);
+
+  const handleClick = useCallback<
+    MouseEventHandler<HTMLButtonElement>
+  >(async () => {
+    if (isPromiseLike(resolvedValue.current)) {
+      setIsLoading(true);
+      resolvedValue.current = await resolvedValue.current;
+      setIsLoading(false);
+    }
+
+    await window.navigator.clipboard.writeText(resolvedValue.current);
+    onCopy?.(resolvedValue.current);
+    setCopied(true);
+
+    timer.current = window.setTimeout(() => {
+      setCopied(false);
+      if (button.current) {
+        button.current.blur();
+      }
+    }, 1000);
+  }, [onCopy]);
 
   return (
-    <div
-      ref={ref}
-      data-test-id={testId}
-      className={cx(styles.wrapper, className)}
-      {...otherProps}
-    >
-      <CopyToClipboard text={value} onCopy={handleOnCopy}>
-        <Tooltip
-          content={copied ? tooltipCopiedText : tooltipText}
-          {...tooltipProps}
-          isDisabled={isDisabled}
-        >
-          <button
-            type="button"
-            ref={button}
-            className={cx(styles.copyButton, {
-              [styles.copyButtonDisabled]: isDisabled,
-            })}
-            aria-label={`Copy ${value} to clipboard`}
-            disabled={isDisabled}
-            aria-live="assertive"
-          >
+    <div ref={ref} data-test-id={testId} className={className} {...otherProps}>
+      <Tooltip
+        content={copied ? tooltipCopiedText : tooltipText}
+        {...tooltipProps}
+        isDisabled={isDisabled}
+      >
+        <Button
+          aria-label={
+            label ??
+            `Copy ${
+              isPromiseLike(resolvedValue.current)
+                ? ''
+                : `"${resolvedValue.current}" `
+            }to clipboard`
+          }
+          aria-live="assertive"
```

Comment:
Voice over isn't working properly. I think we also need to tell it something like (Value "..." has been copied to clipboard) once copied. Currently it just repeats the aria-label value

<img width="725" alt="Screenshot 2023-01-23 at 12 48 33" src="https://user-images.githubusercontent.com/6163988/214035034-827d03d2-83eb-4585-82b2-4a5270456b09.png">

Reply:
Should we move it to the tooltip instead? I think that makes more sense.

Reply:
I decided against that since it's not accessible either. What do you think about the new way I'm doing it? Only problem I'm seeing is that once you copy something it repeats itself and says "Value copied" twice 😅